### PR TITLE
emscripten: 3.1.24 -> 3.1.41, llvmPackages_16: 16.0.1 -> 16.0.6, binaryen: 112 -> 113

### PR DESCRIPTION
**Pull request by willcohen**
###### Description of changes

Emscripten 3.1.39 was able to build from earlier versions, but since emscripten 3.1.39 itself dropped wasm2c support, this should be removed from the derivation.

This plus #235219 should unblock future point releases as well.

Edit: Also includes reverted changes from #229718 and #235219 since my llvm point update was broken and the hashes appeared to have changed upstream as well.

###### Things done

Removed wasm2c from derivation.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
I consider this blocked on `emscripten.src`
https://github.com/NixOS/nixpkgs/pull/229718#issuecomment-1574951156

**Comment by willcohen** [comment]
Apologies, updated hash as noted by #229718!

**Comment by vcunat** [comment]
@ofBorg build emscripten

**Comment by vcunat** [comment]
@ofBorg build
I don't know what's happening, but I'm getting another hash error locally
```
error: hash mismatch in fixed-output derivation '/nix/store/7i99sg5pzmp5kkfbpj85vjrl7n5wziaj-emscripten-node-modules-3.1.39-npm-deps.drv':
         specified: sha256-NSpVXssXwx+94E1qhM3tt2fN2G0EuvPZSN+Xep2IRs8=
            got:    sha256-QlKm6UvPUa7+VJ9ZvXdxYZzK+U96Ju/oAHPhZ/hyv/I=
```

**Comment by willcohen** [comment]
Trying again with the reverted changes from #229718 and #235219 redone here and llvm gnu-install-dirs patch updated. Will run with full `nixpkgs-review` before undrafting.

**Comment by willcohen** [comment]
@vcunat as you predicted, it all came down to the missing patch. should be working now. again, sorry for the need for the reverts.

File: pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch
Comment (willcohen) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: Absolutely. I'd rather avoid a repeat of messing up llvm again, so take as long as you need!

Comment (rrbutani) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: Apologies for the delay.

----

I ended up just adding `${extra_libdir}` back to `_install_rpath`: https://github.com/rrbutani/nixpkgs/commit/666014fe2c731540675c94747431da8b468f8df7; this unbreaks the `x86_64-linux` build. (I haven't tested darwin/aarch64-linux yet but I think it should be fine; if you cherry-pick that commit we can have ofborg give it a go)

I feel a little uneasy adding a patch that has no route to being upstreamed but unfortunately this seems like the least-worst solution; our other options (updating LLVM rdeps to add to `CMAKE_INSTALL_RPATH` either manually or via a setup hook or applying patches to rdeps) all seem too disruptive or brittle.

Despite the discussion on the [differential](https://reviews.llvm.org/D146918) this does _seem_ like a pretty common usecase (depending on a prebuilt/standalone LLVM). For example, other distros do provide LLVM dev packages with `AddLLVM.cmake` and friends (i.e. Ubuntu's [`llvm-15-dev`](https://launchpad.net/ubuntu/+source/llvm-toolchain-15)). It's not apparent to me if/how this patch doesn't break those flows; is it just FHS (assuming the install dir for LLVM and the dependent ends up being some subdir in `/usr/lib`)?

I'd feel better about this if some of the other LLVM nixpkgs maintainers took a look; perhaps I'm missing something obvious.

cc: @alyssais @RaitoBezarius @Ericson2314 @sternenseemann

Comment (RaitoBezarius) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: The patch seems very fair to me, but I do not have more ideas about the ramifications of this whole situation. Though I don't feel strongly about it.

Comment (rrbutani) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: > I was a little disappointed (per discussion) to see this break and tried but failed to argue that the old behaviour should be kept.

Thanks for trying anyways 🙂

> Having a patch to fix this does seem unfortunate if it could be handled by updating `CMAKE_INSTALL_RPATH`. Which seems a big if.

I think ^ would work but I believe we'd have to add LLVM's lib dir to `CMAKE_INSTALL_RPATH` in the build of every LLVM dependent (which seems untenable — without something like a setup hook that does this we'd break out-of-nixpkgs use cases I think).

Does ^ match your understanding? Perhaps I'm missing something.

Comment (willcohen) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: That seems right to me, at least. All of my use cases are very much out of nixpkgs. I suppose we could just tell people that this is what they'd need to add but that seems rude too. Right now I'm leaning toward the patch.

Comment (willcohen) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: Actually, the more I think about it, I feel like the downsides are pretty low. We're already putting in a patch for gnu-install-dirs no matter what, which already broke going from 16.0.1 to 16.0.5 (will soon update to .6), so seems like adding one more change in there isn't the worst thing in the world.

Comment (peterwaller-arm) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: > Still leaning toward the patch, but I suppose switching to the CMAKE variable could be pursued later once it's clearer there's no additional breakage.

I agree with @rrbutani -- the issue is that if LLVM's cmake infrastructure is not patched before it is installed, anything which points itself at the installed LLVM during configure time, when built, will not have the required rpath. The installed LLVM cmake infrastructure ideally should express how to link to the installation at runtime for packages depending on it. Otherwise, users are left expressing how to link it at configure time: If a user wants to develop a package pointing at the LLVM installation, they would also need to set the `CMAKE_INSTALL_RPATH` (since the usual cmake package discovery infrastructure would not be enough to set this).

TL;DR, I can't see a route which isn't patching this at the moment, or having a hook (which would require builds of other packages depending on LLVM to work within the nixpkgs infrastructure to pick up the rpath, I think).

The other ideal situation from nixpkgs point of view (and cmake projects depending on LLVM generally) would be for the upstream patch to be reverted -- but then infrastructure building packages without installing them would need changing to specify that they are doing this (and the build rpath should be dropped).

Comment (rrbutani) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch:55]: > @rrbutani to confirm, the change, were we not to modify `CMAKE_INSTALL_RPATH`, would just be to change the `gnu-install-dirs` patch's revised from ...
> (snipped)

Yup!

I think it'd be good to leave a comment explaining the situation (like [this](https://github.com/willcohen/nixpkgs/compare/emscripten-wasm2c...rrbutani:nixpkgs:llvm-16-bump#diff-9173dc5067b4d545691792b0062164d5e2297e91ac168f9cbadabeb3527493a6R52-R75)), preferably in the actual `AddLLVM.cmake` file in case anyone ends up looking through it to debug stuff.

> We're already putting in a patch for gnu-install-dirs no matter what, which already broke going from 16.0.1 to 16.0.5 (will soon update to .6), so seems like adding one more change in there isn't the worst thing in the world.

I think the difference is that the rest of the `gnu-install-dirs` patch (IIRC — away from a computer so can't double check right now) is in the process of being upstreamed or is at least upstream-able. Agreed that it's not a huge deal though.

**Comment by willcohen** [comment]
@rrbutani Changes applied with latest force push!

**Comment by rrbutani** [comment]
Thanks!

I won't be able to build/test this myself soon but if ofborg is happy I feel good about merging this (the LLVM changes that is); 16.0.6 was a [very small release](https://github.com/llvm/llvm-project/compare/llvmorg-16.0.5...llvmorg-16.0.6).

---

@ofborg build llvmPackages_16

File: pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch
Comment (rrbutani) [reply pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch]: > TL;DR, I can't see a route which isn't patching this at the moment, or having a hook (which would require builds of other packages depending on LLVM to work within the nixpkgs infrastructure to pick up the rpath, I think).

Good point re: having to use the nixpkgs infrastructure; there probably actually _are_ users that use LLVM from nixpkgs in devShells and what not (where the env vars set by setup hooks may not actually be picked up by the actual build tools).

> The other ideal situation from nixpkgs point of view (and cmake projects depending on LLVM generally) would be for the upstream patch to be reverted -- but then infrastructure building packages without installing them would need changing to specify that they are doing this (and the build rpath should be dropped).

IIRC it was mentioned that upstream would be amenable to accepting a build knob for this; maybe we can look into that if keeping this patch up to date proves difficult.

**Comment by willcohen** [comment]
Result of `nixpkgs-review pr 235624` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages marked as broken and skipped:</summary>
  <ul>
    <li>fast-cli</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2jack</li>
    <li>faust2jaqt</li>
    <li>faust2sc</li>
    <li>open-music-kontrollers.mephisto</li>
    <li>puppeteer-cli</li>
    <li>sharedown</li>
    <li>sitespeed-io</li>
    <li>tinygo</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixd</li>
  </ul>
</details>
<details>
  <summary>47 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>clang-tools_16</li>
    <li>clang_16 (llvmPackages_16.clang ,llvmPackages_16.libcxxClang)</li>
    <li>clipboard-jh</li>
    <li>emscripten</li>
    <li>faust (faust2)</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jackrust</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>lld_16 (llvmPackages_16.lld)</li>
    <li>lld_16.dev (llvmPackages_16.lld.dev)</li>
    <li>lld_16.lib (llvmPackages_16.lld.lib)</li>
    <li>lldb_16 (llvmPackages_16.lldb)</li>
    <li>lldb_16.dev (llvmPackages_16.lldb.dev)</li>
    <li>lldb_16.lib (llvmPackages_16.lldb.lib)</li>
    <li>llvmPackages_16.clang-manpages</li>
    <li>llvmPackages_16.libclang (llvmPackages_16.clang-unwrapped)</li>
    <li>llvmPackages_16.libclang.dev (llvmPackages_16.clang-unwrapped.dev)</li>
    <li>llvmPackages_16.libclang.lib (llvmPackages_16.clang-unwrapped.lib)</li>
    <li>llvmPackages_16.libclang.python (llvmPackages_16.clang-unwrapped.python)</li>
    <li>llvmPackages_16.clangNoCompilerRt</li>
    <li>llvmPackages_16.clangNoCompilerRtWithLibc</li>
    <li>llvmPackages_16.clangNoLibc</li>
    <li>llvmPackages_16.clangNoLibcxx</li>
    <li>llvmPackages_16.clangUseLLVM</li>
    <li>llvmPackages_16.compiler-rt (llvmPackages_16.compiler-rt-libc ,llvmPackages_16.compiler-rt-no-libc)</li>
    <li>llvmPackages_16.compiler-rt.dev (llvmPackages_16.compiler-rt-libc.dev)</li>
    <li>llvmPackages_16.libcxx</li>
    <li>llvmPackages_16.libcxx.dev</li>
    <li>llvmPackages_16.stdenv (llvmPackages_16.libcxxStdenv)</li>
    <li>llvmPackages_16.libcxxabi</li>
    <li>llvmPackages_16.libcxxabi.dev</li>
    <li>llvm_16 (llvmPackages_16.llvm)</li>
    <li>llvm_16.dev (llvmPackages_16.llvm.dev)</li>
    <li>llvm_16.lib (llvmPackages_16.llvm.lib)</li>
    <li>llvm_16.python (llvmPackages_16.llvm.python)</li>
    <li>llvmPackages_16.libstdcxxClang</li>
    <li>llvmPackages_16.libunwind</li>
    <li>llvmPackages_16.libunwind.dev</li>
    <li>llvmPackages_16.lldb-manpages</li>
    <li>llvmPackages_16.llvm-manpages</li>
    <li>llvmPackages_16.openmp</li>
    <li>llvmPackages_16.openmp.dev</li>
    <li>vlang</li>
    <li>vtm</li>
  </ul>
</details>

**Comment by willcohen** [comment]
Looks right on Darwin for me — if someone wants to do a nixpkgs-review on Linux could wait for that, otherwise looks like llvmPackages worked okay for ofborg on Linux!

**Review by rrbutani — APPROVED** [review]

**Comment by willcohen** [comment]
@RaitoBezarius @vcunat Any additional questions re llvm, or does this look good?

File: pkgs/development/compilers/emscripten/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/emscripten/default.nix:48]: Please remove the unused fetchpatch from inputs.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/compilers/emscripten/default.nix
Comment (willcohen) [reply pkgs/development/compilers/emscripten/default.nix:48]: @SuperSandro2000 sorry for missing! adjusted with 0761348.

**Comment by rrbutani** [comment]
@willcohen Now that #236420 has hit `master` (as part of the staging cycle in #237233) this will probably have to target staging, unfortunately.

**Comment by willcohen** [comment]
Sounds good! Let me know if any other changes are needed.

**Comment by willcohen** [comment]
Are there any changes anyone would like to see here before merge? From my perspective, it looks ready to go!

**Comment by willcohen** [comment]
Small ping to @NixOS/nixpkgs-committers on this for merge!

**Comment by RaitoBezarius** [comment]
> Small ping to @NixOS/nixpkgs-committers on this for merge!

Done :)

**Comment by vcunat** [comment]
Well, be aware that the group has around 200 members...

**Pull request merged**
